src/lsm/manifest.zig

```diff
@@ -222,11 +225,46 @@ pub fn ManifestType(comptime Table: type, comptime Storage: type) type {
             manifest_level.set_snapshot_max(snapshot, table);
             assert(table.snapshot_max == snapshot);
 
-            // Append update changes to the manifest log
+            // Append update changes to the manifest log.
             const log_level = @intCast(u7, level);
             manifest.manifest_log.insert(log_level, table);
         }
 
+        pub fn move_table(
+            manifest: *Manifest,
+            level_a: u8,
+            level_b: u8,
+            snapshot: u64,
+            table: *TableInfo,
+        ) void {
+            assert(level_b == level_a + 1);
+            assert(level_b < constants.lsm_levels);
+            assert(table.snapshot_max >= snapshot);
+
+            const manifest_level_a = &manifest.levels[level_a];
+            const manifest_level_b = &manifest.levels[level_b];
+            if (constants.verify) {
+                assert(manifest_level_a.contains(table));
+                assert(!manifest_level_b.contains(table));
+            }
+
+            // Insert the table into level b and append insert changes to the manifest log.
+            manifest_level_b.insert_table(manifest.node_pool, table);
+            manifest.manifest_log.insert(@intCast(u7, level_b), table);
+
+            // Update the table in level a without appending update changes to the manifest log.
+            // To move a table w.r.t manifest log, the previous level's table should not be removed.
+            // When replaying the log from open(), inserts are processed in LIFO order while
+            // duplicates are ignored. This means the table will only appear in level_b as moved.
+            manifest_level_a.set_snapshot_max(snapshot, table);
```

Comment:
Is it possible to hoist this above the "insert to level_b"? So we deal with level A and then level B in order?

Is there a bug here where we might have two copies of the table in both levels with the same snapshot? CC'ing @sentientwaffle

Reply:
Currently the order is important because we need to ensure that `manifest_level_b.insert_table` happens before `manifest_level_a.set_snapshot_max` modifies the table's snapshot.

But that is subtle — maybe we should instead copy `table` to the stack (`s/table/table_a` + `const table_b = table_a.*`), then handle them in order (`manifest_level_a.set_snapshot_max(snapshot, table_a);` then `manifest_level_b.insert_table(manifest.node_pool, table_b);`.

Reply:
> Also, is it possible that the two tables share the same snapshot value, to be seen as duplicates (if level b was also probed in the read path by a subsequent snapshot)?

Just to double-check, you mean that a subsequent query overlaps with the table in both level A and B?
Maybe we should remove the table from `manifest_level_a` instead of just setting its snapshot. I think that would also negate the need for `moved_table_addresses`.

Reply:
Would removing from `manifest_level_a` during `move_table` still invalidate a subsequent `StateMachine.prefetch` in the middle of a half-measure? https://github.com/tigerbeetledb/tigerbeetle/pull/463#discussion_r1113459688

Reply:
By avoiding the snapshot update I think we would avoid triggering `remove_invisible_tables`.

Reply:
It would be called on level A at the end (since the Compaction was started), but it wont have anything to remove. IIUC, the primary difference would be that, after a tree.compact() in the middle of a half-measure, prefetch() can observe that the table either still exists in level A with the updated snapshot or doesn't exist in level A at all but atomically now exists in level B.

Reply:
> prefetch() can observe that the table either still exists in level A with the updated snapshot or doesn't exist in level A at all but atomically now exists in level B.

Your wording here confuses me. My thinking is:

- When we do a move-table, we don't update the snapshot at all.
- We atomically move the table (in memory, i.e. `ManifestLevel`s) from level A to level B.

After the atomic move, the table is still available, just from a different level.

Reply:
> After the atomic move, the table is still available, just from a different level.

Yes, this feels like an easier mental model to me. I can't think of any counter examples where it could go wrong.